pygmt/datatypes.py

```diff
@@ -0,0 +1,221 @@
+        ("ProjRefEPSG", ctp.c_int),
+        # Book-keeping variables "hidden" from the API
+        ("hidden", ctp.c_void_p),
+    ]
+
+    def to_vectors(self):
+        """
+        Convert a GMT_DATASET object to a list of vectors.
+
+        Currently, the number of columns in all segments of all tables are
+        assumed to be the same. The same column in all segments of all tables
+        are concatenated. The trailing text column is also concatenated as a
+        string vector.
+
+        Examples
+        --------
+        >>> from pygmt.helpers import GMTTempFile
+        >>> from pygmt.clib import Session
+        >>>
+        >>> with GMTTempFile(suffix=".txt") as tmpfile:
+        ...     # prepare the sample data file
+        ...     with open(tmpfile.name, mode="w") as fp:
+        ...         print(">", file=fp)
+        ...         print("1.0 2.0 3.0 TEXT1 TEXT23", file=fp)
+        ...         print("4.0 5.0 6.0 TEXT4 TEXT567", file=fp)
+        ...         print(">", file=fp)
+        ...         print("7.0 8.0 9.0 TEXT8 TEXT90", file=fp)
+        ...         print("10.0 11.0 12.0 TEXT123 TEXT456789", file=fp)
+        ...     with Session() as lib:
+        ...         with lib.virtualfile_to_data(kind="dataset") as vouttbl:
+        ...             lib.call_module(
+        ...                 "read", f"{tmpfile.name} {vouttbl} -Td"
+        ...             )
+        ...             ds = lib.read_virtualfile(vouttbl, kind="dataset")
+        ...             vectors = ds.contents.to_vectors()
+        ...
+        >>> len(vectors)  # 4 columns
+        4
+        >>> vectors[0]
+        array([ 1.,  4.,  7., 10.])
+        >>> vectors[1]
+        array([ 2.,  5.,  8., 11.])
+        >>> vectors[2]
+        array([ 3.,  6.,  9., 12.])
+        >>> vectors[3]
+        array(['TEXT1 TEXT23', 'TEXT4 TEXT567', 'TEXT8 TEXT90',
+                'TEXT123 TEXT456789'], dtype='<U18')
+
+        Returns
+        -------
+        vectors : list of 1-D arrays
+            List of vectors containing the data from the GMT_DATASET object.
+        """
+        vectors = []
+        for icol in range(self.n_columns):
+            colvector = []
+            for itbl in range(self.n_tables):
+                dtbl = self.table[itbl].contents
+                for iseg in range(dtbl.n_segments):
+                    dseg = dtbl.segment[iseg].contents
+                    colvector.append(
+                        np.ctypeslib.as_array(dseg.data[icol], shape=(dseg.n_rows,))
+                    )
+            vectors.append(np.concatenate(colvector))
+
+        # deal with trailing text column
+        textvector = []
+        for itbl in range(self.n_tables):
+            dtbl = self.table[itbl].contents
+            for iseg in range(dtbl.n_segments):
+                dseg = dtbl.segment[iseg].contents
+                if dseg.text:
+                    textvector.extend(dseg.text[: dseg.n_rows])
```

Comment:
Great suggestion! 

In commit 87424677f37158be872c42c344ed02c0aa68a261, I modified the old `to_vectors()` method to the new `to_dataframe()` method, which returns a pd.DataFrame object instead. The `pd.DataFrame` object is built using [`pd.concat`](https://pandas.pydata.org/docs/reference/api/pandas.concat.html), rather than `pd.DataFrame.from_dict(dict(zip(colnames, vectors)))`, which is more elegant.